### PR TITLE
Mutation, Composition, and Utility Generators

### PR DESCRIPTION
**Pull request by Razaekel**
Opening this WIP PR so I can get feedback on the implementation as I work on it.

Reference Issue: #100 

Currently, I've only added the Checkerboard and Constant Generator modules while I worked out the type system and implementation. It ended up similar to Aatch's ValueSourceN trait, without the N portion as I pass that though the function implementation instead. I have not figured out how to implement Aatch's ValueSource with associated Input and Output types yet. Hints would be helpful, since this would make it much simpler to implement dimension-agnostic modules like Select or Constant, as Aatch proposed.

My intention is to give every generator/modifier/mutator the Module trait, thereby allowing composition through the use of Modules.

The foundation of all the modules I plan to implement is based on libnoise.

For a minimal implementation, I'll add the following modules:
- [x] Checkerboard
- [x] Constant
- [x] Cylinders
- [x] Perlin
- [x] Select

Modules to be included in later PRs:
Simple Generators:
- OpenSimplex
- Simplex (2D only, 3D and up is covered by patent)
- Spheres
- Voronoi

Fractal Generators:
- Billow
- Multifractal
- Ridged Multifractal

Combiners:
- Add
- Max
- Min
- Multiply
- Power

Modifier Modules:
- Abs
- Clamp
- Curve
- Exponent
- Invert
- ScaleBias
- Terrace

Selector Modules:
- Blend

Transformer Modules:
- Displace
- RotatePoint
- ScalePoint
- TransformPoint
- Turbulence

Miscellaneous Modules:
- Cache

I'm open to other module suggestions, as well. There's probably a bunch more that'd fit under Fractal Generators, since there's a lot of ways to combine perlin noise in octaves.

### CODE REVIEW

**Comment by bjz** [comment]
cc @Aatch

**Comment by Razaekel** [comment]
I'm definitely going to need to figure out how to get a dimension-agnostic type system working, since after thinking about it, the following modules don't care about the dimension, since they only work with the value. It would suck to have to write out the same functions for every dimension in these modules, instead of making them generic.

- Constant
- All Combiner Modules
- All Modifier Modules
- All Selector Modules
- Cache

The generators (except Constant), and the transformers care about the dimension of the point, since they either use the point or they modify the point.

File: src/modules/generators/checkerboard.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+// limitations under the License.
+
+use num::Float;
+use math;
+use Module;
+use math::{Point2, Point3};
+
+pub struct Checkerboard {
+	size: usize
+}
```
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: Seems like you have tabs set in your editor - standard for Rust is four spaces.

File: src/modules/generators/checkerboard.rs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+	fn get(&self, point: Point2<T>) -> Self::Output {
+		let result = point.iter()
+			.map(|&a| fastfloor(a))
+			.fold(0, |a, b| (a & self.size) ^ (b & self.size));
+
+		if result > 0 {
+			math::cast::<_, T>(-1.0)
+		} else {
+		    math::cast::<_, T>(1.0)
```
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: Rust should be able to infer the casts, no?

File: src/modules/generators/cylinders.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	let result = ((x0 & size) ^ (y0 & size));
+	if (result > 0) {
+		value = -1.0;
+	}
+	else {
+	    value = 1.0;
+	}
+
+	math::cast::<_, T>(value)
+}
```
Comment (bjz) [inline src/modules/generators/cylinders.rs]: You should also set your editor and git to add a new line at the end of files

**Comment by bjz** [comment]
As a tip: import `num::Zero` or `num::One`, and then you can do `T::zero()` or `T::one()` respectively. I kind of prefer that to `cast::<_, T>(0.0)`.

File: src/modules/generators/checkerboard.rs (hunk 2 above)
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: See comment about `T::one()`

File: src/lib.rs [hunk 4]
```diff
@@ -24,7 +24,7 @@
 //! let val = noise.apply(&seed, &[42.0, 37.0, 2.0]);
 //! ```
 
-#![deny(missing_copy_implementations)]
+//#![deny(missing_copy_implementations)]
```
Comment (bjz) [inline src/lib.rs]: Why was this removed?

File: src/modules/generators/checkerboard.rs (hunk 1 above)
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: Adding a `#[derive(Copy, Clone)]` here will solve the `#![deny(missing_copy_implementations)]` issue I mentioned before.

File: src/modules/generators/cylinders.rs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use num::Float;
+use {math, lib};
+
+pub struct Cylinders<T> {
+	frequency: T,
+}
```
Comment (bjz) [inline src/modules/generators/cylinders.rs]: I wonder if calling it `Cylinder` would be more consistent?

Comment (Razaekel) [inline src/modules/generators/cylinders.rs]: Not really, the implementation creates concentric cylinders, not a single cylinder.

File: src/lib.rs (hunk 4 above)
Comment (Razaekel) [inline src/lib.rs]: It was giving me an error about missing copy implementations on the Checkerboard struct. Fixed it in 9884435.

**Comment by Razaekel** [comment]
Major thanks to @Aatch for helping me figure out the type signature to use for the Select module.

The select example shows blending between the two source modules, and uses the checkerboard as the control module.

File: examples/cylinders.rs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+            pixels.push(cast(clamp(value * 0.5 + 0.5, 0.0, 1.0) * 255.0));
+        }
+    }
+
+    let _ = image::save_buffer(&Path::new(filename),
+                               &*pixels,
+                               width,
+                               height,
+                               image::Gray(8));
+}
```
Comment (bjz) [inline examples/cylinders.rs]: The examples probably don't need to be generic - `f32` or `f64` should be fine! :)

File: src/modules/generators/checkerboard.rs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+impl Checkerboard {
+    pub fn new(s: usize) -> Checkerboard {
+        Checkerboard { size: 1 << s }
+    }
+}
+
+fn fastfloor<T: Float>(x: T) -> usize {
+    if x > math::cast::<_, T>(0) {
```
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: Could be `T::zero()`

File: src/modules/generators/checkerboard.rs [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        math::cast::<_, usize>(x)
```
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: Dropping the explicit type param work?

File: src/modules/generators/cylinders.rs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    type Output = T;
+
+    fn get(&self, point: Point3<T>) -> Self::Output {
+        let x = point[0] * self.frequency;
+        let z = point[2] * self.frequency;
+
+        let dist_from_center: T = (x * x + z * z).sqrt();
+        let dist_from_smaller_sphere: T = dist_from_center - dist_from_center.floor();
+        let dist_from_larger_sphere: T = T::one() - dist_from_smaller_sphere;
+        let nearest_dist: T = dist_from_smaller_sphere.min(dist_from_larger_sphere);
```
Comment (bjz) [inline src/modules/generators/cylinders.rs]: Probably no need for explicit types on these bindings.

File: src/modules/generators/checkerboard.rs [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    }
+}
+
+impl<T: Float> Module<Point2<T>> for Checkerboard {
+    type Output = T;
+
+    fn get(&self, point: Point2<T>) -> Self::Output {
+        let result = point.iter()
+            .map(|&a| fastfloor(a))
+            .fold(0, |a, b| (a & self.size) ^ (b & self.size));
```
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: `math::map2` then `math::fold2`?

File: src/modules/generators/checkerboard.rs [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    }
+}
+
+impl<T: Float> Module<Point3<T>> for Checkerboard {
+    type Output = T;
+
+    fn get(&self, point: Point3<T>) -> Self::Output {
+        let result = point.iter()
+            .map(|&a| fastfloor(a))
+            .fold(0, |a, b| (a & self.size) ^ (b & self.size));
```
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: `math::map1` then `math::fold1`?

File: src/modules/generators/checkerboard.rs [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+        if result > 0 {
+            -T::one()
+        } else {
+            T::one()
+        }
+    }
+}
```
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: What about an impl for `Point4`?

File: src/lib.rs [hunk 13]
```diff
@@ -95,3 +97,9 @@ pub trait GenFn4<T>: Fn(&Seed, &Point4<T>) -> T {}
 impl<T, F> GenFn2<T> for F where F: Fn(&Seed, &Point2<T>) -> T {}
 impl<T, F> GenFn3<T> for F where F: Fn(&Seed, &Point3<T>) -> T {}
 impl<T, F> GenFn4<T> for F where F: Fn(&Seed, &Point4<T>) -> T {}
+
+pub trait Module<T> {
```
Comment (bjz) [inline src/lib.rs]: I wonder if `Source` or `ValueSource` would be a better term? `Module` seems rather generic a term. @Aatch?

File: src/modules/mod.rs [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub use self::generators::*;
+pub use self::selectors::*;
+
+mod generators;
+mod selectors;
```
Comment (bjz) [inline src/modules/mod.rs]: Missing new line at the end of this file.

File: src/modules/selectors/mod.rs [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub use self::select::*;
+
+mod select;
```
Comment (bjz) [inline src/modules/selectors/mod.rs]: Missing new line at the end of this file.

File: src/modules/selectors/select.rs [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+/// Maps a value onto a cubic S-curve.
+fn s_curve3<T: Float>(a: T) -> T {
+    a * a * (math::cast::<_, T>(3.0) - (math::cast::<_, T>(2.0) * a))
+}
+
+/// Performs linear interploation between two values.
+fn linear_interp<T: Float>(lhs: T, rhs: T, a: T) -> T {
+    ((T::one() - a) * lhs) + (a * rhs)
+}
```
Comment (bjz) [inline src/modules/selectors/select.rs]: Maybe chuck this in the `math` module?

File: examples/constant.rs [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+use noise::modules::Constant;
+
+mod debug;
+
+fn main() {
+    debug::render_png2("constant1.png", Constant::new(-1.0), 1024, 1024, 1);
+    debug::render_png2("constant2.png", Constant::new(0.0), 1024, 1024, 1);
+    debug::render_png2("constant3.png", Constant::new(1.0), 1024, 1024, 1);
+
+    println!("Generated constant1.png, constant2.png, and constant3.png");
```
Comment (bjz) [inline examples/constant.rs]: Maybe move this logging to the end of the `render_png2` and `render_png` functions. That way you get one for each file as it processes.

File: src/modules/selectors/select.rs (hunk 16 above)
Comment (Razaekel) [inline src/modules/selectors/select.rs]: I'll move them to a separate interp module, so I can put other interpolation methods in there as well.

File: src/lib.rs (hunk 13 above)
Comment (Razaekel) [inline src/lib.rs]: I don't know about changing the name of Module to Source/ValueSource. Not all of these will 'source' a value. Some will modify their passed values, others won't modify the value, but modify the point instead.

It'd be nice to sort them as SourceModule, CombinerModule, ModifierModule, etc., but that's probably being too specific.

File: src/modules/generators/checkerboard.rs (hunk 10 above)
Comment (Razaekel) [inline src/modules/generators/checkerboard.rs]: I'm keeping the implementation the same between the different PointNs so that at some later point when we have a higher Point type, they can be made generic over Point instead of Point2/3/N.

File: examples/constant.rs (hunk 17 above)
Comment (Razaekel) [inline examples/constant.rs]: Done in 44689b

File: src/modules/generators/checkerboard.rs (hunk 12 above)
Comment (Razaekel) [inline src/modules/generators/checkerboard.rs]: Added in 0628827.

**Comment by Razaekel** [comment]
At this point, this PR is no longer WIP.

File: src/modules/generators/cylinders.rs [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+        let x = point[0] * self.frequency;
+        let z = point[2] * self.frequency;
+
+        let dist_from_center = (x * x + z * z).sqrt();
+        let dist_from_smaller_sphere = dist_from_center - dist_from_center.floor();
+        let dist_from_larger_sphere = T::one() - dist_from_smaller_sphere;
+        let nearest_dist = dist_from_smaller_sphere.min(dist_from_larger_sphere);
+        T::one() - (nearest_dist * math::cast(4.0))
+    }
+}
```
Comment (bjz) [inline src/modules/generators/cylinders.rs]: So it is only possible to implement cylinders for 3d points?

File: src/modules/selectors/select.rs [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+
+    fn get(&self, point: T) -> Self::Output {
+        let control_value = self.selector.get(point);
+
+        if self.edge_falloff > U::zero() {
+            if control_value < (self.lower_bound - self.edge_falloff) {
+                self.source1.get(point)
+            } else if control_value < (self.lower_bound + self.edge_falloff) {
+                let lower_curve: U = self.lower_bound - self.edge_falloff;
+                let upper_curve: U = self.lower_bound + self.edge_falloff;
```
Comment (bjz) [inline src/modules/selectors/select.rs]: No need for these type annotations :)

File: src/interp.rs [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+/// Performs linear interploation between two values.
+pub fn linear<T: Float>(lhs: T, rhs: T, a: T) -> T {
+    ((T::one() - a) * lhs) + (a * rhs)
+}
+
+/// Maps a value onto a cubic S-curve.
+pub fn s_curve3<T: Float>(a: T) -> T {
+    a * a * (math::cast::<_, T>(3.0) - (math::cast::<_, T>(2.0) * a))
+}
+
```
Comment (bjz) [inline src/interp.rs]: I dunno - still feel like it would be ok to put these in `math` for now, as opposed to prematurely putting them in a different module.

File: src/modules/selectors/select.rs [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+                let alpha = interp::s_curve3((control_value - lower_curve) / (upper_curve - lower_curve));
+
+                interp::linear(self.source1.get(point), self.source2.get(point), alpha)
+            } else if control_value < (self.upper_bound - self.edge_falloff) {
+                self.source2.get(point)
+            } else if control_value < (self.upper_bound + self.edge_falloff) {
+                let lower_curve: U = self.upper_bound - self.edge_falloff;
+                let upper_curve: U = self.upper_bound + self.edge_falloff;
```
Comment (bjz) [inline src/modules/selectors/select.rs]: ditto

File: src/modules/selectors/select.rs [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+/// - Source Module 1 outputs a value.
+/// - Source Module 2 outputs a value.
+/// - Control Module determines the value to select. If the output value
+/// from the control module is within a range of values known as the selection
+/// range, this noise module outputs the value from Source Module 2. Otherwise,
+/// this noise module outputs the value from Source Module 1.
+pub struct Select<Lhs, Rhs, Control, U> {
+    source1: Lhs,
+    source2: Rhs,
+    selector: Control,
```
Comment (bjz) [inline src/modules/selectors/select.rs]: Perhaps call this field `control`?

File: src/modules/selectors/select.rs [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+    edge_falloff: U,
+    lower_bound: U,
+    upper_bound: U,
+}
+
+impl<Lhs, Rhs, Control, U> Select<Lhs, Rhs, Control, U> {
+    pub fn new(s1: Lhs, s2: Rhs, c: Control, falloff: U, lower: U, upper: U) -> Select<Lhs, Rhs, Control, U> {
```
Comment (bjz) [inline src/modules/selectors/select.rs]: Spell these identifiers out fully instead of shortening them.

File: src/modules/selectors/select.rs [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+/// from the control module is within a range of values known as the selection
+/// range, this noise module outputs the value from Source Module 2. Otherwise,
+/// this noise module outputs the value from Source Module 1.
+pub struct Select<Lhs, Rhs, Control, U> {
+    source1: Lhs,
+    source2: Rhs,
+    selector: Control,
+    edge_falloff: U,
+    lower_bound: U,
+    upper_bound: U,
```
Comment (bjz) [inline src/modules/selectors/select.rs]: You could make all these fields public, then document them individually with doc comments. Same for the other modules.

File: src/modules/generators/perlin.rs [hunk 25]
```diff
@@ -0,0 +1,211 @@
...
+use math::{Point2, Point3, Point4};
+
+use {gradient, Seed};
+
+pub struct Perlin {
+    seed: Seed,
+}
+
+impl Perlin {
+    pub fn new(s: Seed) -> Perlin {
```
Comment (bjz) [inline src/modules/generators/perlin.rs]: Spell this param out fully.

File: src/modules/generators/cylinders.rs [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+use math::Point3;
+
+use Module;
+
+pub struct Cylinders<T: Float> {
+    frequency: T,
+}
+
+impl<T: Float> Cylinders<T> {
+    pub fn new(v: T) -> Cylinders<T> {
```
Comment (bjz) [inline src/modules/generators/cylinders.rs]: Spell this out fully.

File: src/modules/generators/cylinders.rs [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+
+use num::{Float, One};
+use math;
+use math::Point3;
+
+use Module;
+
+pub struct Cylinders<T: Float> {
+    frequency: T,
+}
```
Comment (bjz) [inline src/modules/generators/cylinders.rs]: Add some documentation to the struct and the field? And make the field public?

File: src/modules/generators/perlin.rs [hunk 28]
```diff
@@ -0,0 +1,211 @@
...
+use num::{Float, One};
+use math;
+use Module;
+use math::{Point2, Point3, Point4};
+
+use {gradient, Seed};
+
+pub struct Perlin {
+    seed: Seed,
+}
```
Comment (bjz) [inline src/modules/generators/perlin.rs]: Documentation, and make the field public?

File: src/modules/generators/constant.rs [hunk 29]
```diff
@@ -0,0 +1,52 @@
...
+
+pub struct Constant<T: Float> {
+    value: T,
+}
+
+impl<T: Float> Constant<T> {
+    pub fn new(v: T) -> Constant<T> {
+        Constant { value: v }
+    }
+}
```
Comment (bjz) [inline src/modules/generators/constant.rs]: Same comments as with the other modules.

File: src/modules/generators/checkerboard.rs [hunk 30]
```diff
@@ -0,0 +1,87 @@
...
+#[derive(Clone,Copy,Debug)]
+pub struct Checkerboard {
+    size: usize,
+}
+
+impl Checkerboard {
+    pub fn new(s: usize) -> Checkerboard {
+        Checkerboard { size: 1 << s }
+    }
+}
```
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: Same comments as with the other modules

File: src/modules/generators/checkerboard.rs [hunk 31]
```diff
@@ -0,0 +1,87 @@
...
+
+fn fastfloor<T: Float>(x: T) -> usize {
+    if x > T::zero() {
+        math::cast(x)
+    } else {
+        math::cast(x - T::one())
+    }
+}
```
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: Make this `fast_floor` and move it to `math`?

File: src/modules/generators/checkerboard.rs (hunk 30 above)
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: Why is the shift needed? Could we add a comment for that?

File: src/modules/generators/checkerboard.rs [hunk 32]
```diff
@@ -0,0 +1,87 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use num::{Float, One};
+use math;
+use Module;
+use math::{Point2, Point3, Point4};
+
+#[derive(Clone,Copy,Debug)]
```
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: Add spacing between the commas. Could you ensure that the other modules have the same derives added to them?

File: src/lib.rs (hunk 13 above)
Comment (bjz) [inline src/lib.rs]: Hmm... I'm actually thinking that these are really just functions. Perhaps we could use my `https://crates.io/crates/fn_ops` crate? We would then be implementing [`Fn<(T,)>`](http://bjz.github.io/fn_ops/fn_ops/trait.Fn.html) (see the [example](http://bjz.github.io/fn_ops/fn_ops/index.html#example)). The cool thing would be that you could then pass in functions or closures to the generators as well!

File: src/modules/generators/checkerboard.rs (hunk 32 above)
Comment (Razaekel) [inline src/modules/generators/checkerboard.rs]: I can't put Copy on the Perlin module because Seed does not implement it.

File: src/modules/selectors/select.rs (hunk 24 above)
Comment (Razaekel) [inline src/modules/selectors/select.rs]: My intent was to keep the fields immutable, since the struct is public. I'd like to eventually making this library multi-threading compatable, and allowing the modules to be mutable could affect the consistency of results.

File: src/interp.rs (hunk 20 above)
Comment (Razaekel) [inline src/interp.rs]: Would a interp submodule in math be okay? I like the interp::linear syntax.

File: src/modules/generators/checkerboard.rs (hunk 30 above)
Comment (Razaekel) [inline src/modules/generators/checkerboard.rs]: I'll add a comment, but size needs to be a power of 2 to create a correct checkerboard. Non-power of 2 sizes produces irregular boards.

File: src/modules/generators/checkerboard.rs (hunk 32 above)
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: Well, just the ones that you can. Debug at least would be great!

File: src/modules/generators/perlin.rs [hunk 33]
```diff
@@ -0,0 +1,211 @@
...
+                             -> T {
+            let attn = T::one() - math::dot2(distance, distance);
+            if attn > T::zero() {
+                math::pow4(attn) * math::dot2(distance, gradient::get2(seed.get2(corner)))
+            } else {
+                T::zero()
+            }
+        }
+
+        let floored = math::map2(point, Float::floor);
```
Comment (bjz) [inline src/modules/generators/perlin.rs]: I'd prefer `T::floor` for this.

Comment (Razaekel) [inline src/modules/generators/perlin.rs]: The perlin code was a straight copy of the code in src/perlin.rs.

File: src/modules/generators/checkerboard.rs (hunk 32 above)
Comment (Razaekel) [inline src/modules/generators/checkerboard.rs]: Seed doesn't implement any of those traits, so I can't implement it on structs that include Seed. I can add it on after it has been added onto Seed.

**Comment by Razaekel** [comment]
On further consideration, this should remain a WIP until all the little details are ironed out and I have a chance to squash all the commits down to a smaller set.

**Comment by Razaekel** [comment]
Rebased to current master, and modified select to take references so that the input modules can be reused.

File: src/value.rs [hunk 34]
```diff
@@ -13,22 +13,9 @@
...
-    a + x * (b - a)
-}
-
-/// Map value between `0.0` and `1.0` to a Quintic Hermite curve.
-fn smoothstep<T: Float>(x: T) -> T {
-    let _15: T = math::cast(15.0);
-    let _10: T = math::cast(10.0);
-    let _6: T = math::cast(6.0);
-    x * x * x * (x * (x * _6 - _15) + _10)
-}
```
Comment (bjz) [inline src/value.rs]: 👍

**Comment by Razaekel** [comment]
I can now consider this no longer a WIP. The minimal implementation of the proposed API discussed in #100 is completed, and functions as I want it to in regards to generating and composing the noise modules.

File: src/math.rs [hunk 35]
```diff
@@ -216,3 +216,23 @@ pub fn cast4<T, U>(x: Point4<T>) -> Point4<U>
...
     map4(x, cast)
 }
+
+pub mod interp {
+    use num_traits::Float;
+    use math;
+
+    /// Performs linear interploation between two values.
+    pub fn linear<T: Float>(a: T, b: T, x: T) -> T {
+        a + x * (b - a)
```
Comment (bjz) [inline src/math.rs]: I wonder if we can use `mul_add` for this to improve performance?

File: src/modules/combiners/select.rs [hunk 36]
```diff
@@ -0,0 +1,102 @@
...
+///
+/// - Source Module 1 outputs a value.
+/// - Source Module 2 outputs a value.
+/// - Control Module determines the value to select. If the output value
+/// from the control module is within a range of values known as the selection
+/// range, this noise module outputs the value from Source Module 2. Otherwise,
+/// this noise module outputs the value from Source Module 1.
+pub struct Select<'a, Lhs: 'a, Rhs: 'a, Control: 'a, U> {
+    source1: &'a Lhs,
+    source2: &'a Rhs,
```
Comment (bjz) [inline src/modules/combiners/select.rs]: Could we make these names consistent with the type parameters?

File: src/modules/combiners/select.rs [hunk 37]
```diff
@@ -0,0 +1,102 @@
...
+    selector: &'a Control,
+    edge_falloff: U,
+    lower_bound: U,
+    upper_bound: U,
+}
+
+impl<'a, Lhs, Rhs, Control, U> Select<'a, Lhs, Rhs, Control, U> {
+    pub fn new(s1: &'a Lhs,
+               s2: &'a Rhs,
+               c: &'a Control,
```
Comment (bjz) [inline src/modules/combiners/select.rs]: Would be cool to spell out these parameters fully.

File: src/modules/generators/checkerboard.rs (hunk 32 above)
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: Another ping to fix the spacing.

File: src/lib.rs [hunk 38]
```diff
@@ -97,3 +99,9 @@ pub trait GenFn4<T>: Fn(&PermutationTable, &Point4<T>) -> T {}
 impl<T, F> GenFn2<T> for F where F: Fn(&PermutationTable, &Point2<T>) -> T, {}
 impl<T, F> GenFn3<T> for F where F: Fn(&PermutationTable, &Point3<T>) -> T, {}
 impl<T, F> GenFn4<T> for F where F: Fn(&PermutationTable, &Point4<T>) -> T, {}
+
+pub trait Module<T> {
+    type Output;
+
+    fn get(&self, point: T) -> Self::Output;
+}
```
Comment (bjz) [inline src/lib.rs]: If we add this:

```rust
impl<'a, T, M: Module<T>> Module<T> for &'a M {
    type Output = M::Output;

    #[inline]
    fn get(&self, point: T) -> M::Output {
        M::get(*self, point)
    }
}
```

`Select` wouldn't have to specify whether its modules are taken by value or by ref.

File: src/lib.rs [hunk 39]
```diff
@@ -97,3 +99,9 @@ pub trait GenFn4<T>: Fn(&PermutationTable, &Point4<T>) -> T {}
 impl<T, F> GenFn2<T> for F where F: Fn(&PermutationTable, &Point2<T>) -> T, {}
 impl<T, F> GenFn3<T> for F where F: Fn(&PermutationTable, &Point3<T>) -> T, {}
 impl<T, F> GenFn4<T> for F where F: Fn(&PermutationTable, &Point4<T>) -> T, {}
+
+pub trait Module<T> {
```
Comment (bjz) [inline src/lib.rs]: I would still prefer this to be called `ValueSource` or something. Also, we need docs for this trait.

File: src/math.rs [hunk 40]
```diff
@@ -216,3 +216,23 @@ pub fn cast4<T, U>(x: Point4<T>) -> Point4<U>
...
+    }
+
+    /// Maps a value onto a cubic S-curve.
+    pub fn s_curve3<T: Float>(x: T) -> T {
```
Comment (bjz) [inline src/math.rs]: What about `sigmoid3`? A link to https://en.wikipedia.org/wiki/Sigmoid_function might be cool too.

File: src/modules/generators/checkerboard.rs (hunk 30 above)
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: We should give the argument name, and add a doc comment specifying what it does.

File: src/modules/generators/checkerboard.rs (hunk 11 above)
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: I would prefer these be used in the interim, to be consistent with other parts of the codebase.

File: src/modules/generators/cylinders.rs (hunk 27 above)
Comment (Razaekel) [inline src/modules/generators/cylinders.rs]: I was intending for the fields to remain private so that the modules would stay immutable once generated, preventing the use of mut Module to modify the internal field values.

File: src/lib.rs (hunk 39 above)
Comment (Razaekel) [inline src/lib.rs]: Still not sure about changing the name of Module to Source/ValueSource. Not all of these will 'source' a value. Some will modify their passed values, others won't modify the value, but modify the point instead.

A quick google search indicates that most of the other language implementations of procedural noise are using either Module, IModule, or ModuleBase as the common parent name.

File: src/lib.rs (hunk 38 above)
Comment (Razaekel) [inline src/lib.rs]: Wouldn't it be helpful to require that modules be taken by ref, instead of leaving open the possibility of taking them by value, thereby leaving open a possible source of bugs? or should we just expect users to get corrected by rustc where applicable?

File: src/modules/generators/checkerboard.rs (hunk 11 above)
Comment (Razaekel) [inline src/modules/generators/checkerboard.rs]: I can do the initial map2 on this, but I can't use fold2 on it because of the '& self.size'. Adding another map2 to apply the '& self.size' gives me "binary operation & cannot be applied to type 'T'". Even after that, fold gives me "binary operation '^' cannot be applied to type '[usize; 2]'"

I'm going to have to leave it as it is for now, since it works with the iterator.

File: src/lib.rs (hunk 39 above)
Comment (Razaekel) [inline src/lib.rs]: How about NoiseModule?

File: src/modules/generators/checkerboard.rs (hunk 30 above)
Comment (Razaekel) [inline src/modules/generators/checkerboard.rs]: Added.

File: src/lib.rs (hunk 38 above)
Comment (bjz) [inline src/lib.rs]: We have a type system for that ^_^

File: src/modules/generators/checkerboard.rs (hunk 30 above)
Comment (bjz) [inline src/modules/generators/checkerboard.rs]: The argument name to the constructor is still `s`

Comment (Razaekel) [inline src/modules/generators/checkerboard.rs]: whoops

File: src/lib.rs (hunk 38 above)
Comment (Razaekel) [inline src/lib.rs]: Added. This let me remove the lifetime specifiers from Select, and remove some of the refs from the select example. The Perlin module still had to be specified by ref since it doesn't have a copy implementation.

Comment (Razaekel) [inline src/lib.rs]: Actually, with the Clone impl you added to PermutationTable, I can add a #[derive(Copy)] to it and now Perlin can derive Copy as well.

This leads to the question of whether it's better to emphasize using references instead of copy/move, since the Perlin module could take up a decent bit of memory, or more specifically any module that contains a PermutationTable. Long noise chains would inflate the amount of memory used if they use the same PermutationTable-containing module multiple times.

Comment (Razaekel) [inline src/lib.rs]: Eh, I guess I'm just bikeshedding here.

File: src/permutationtable.rs [hunk 41]
```diff
@@ -26,7 +26,7 @@ const TABLE_SIZE: usize = 256;
 ///
 /// Table creation is expensive, so in most circumstances you'll only want to
 /// create one of these per generator.
-#[allow(missing_copy_implementations)]
+#[derive(Copy)]
```
Comment (bjz) [inline src/permutationtable.rs]: The reason I didn't derive `Copy` was because the table was so large.... maybe you are right about defaulting to references... I dunno. >_<

**Comment by bjz** [comment]
I would like to merge this asap. It's been open for far too long - probably my own fault... wondering if we could put this behind a feature flag for now - like `experimental` or something...

File: src/permutationtable.rs (hunk 41 above)
Comment (Razaekel) [inline src/permutationtable.rs]: I started a branch off of this that removes the Copy on PermutationTable [here](/Razaekel/noise-rs/tree/NoCopyPermutationTable). I haven't gone so far as to require only references on Select yet.

**Comment by Razaekel** [comment]
Changed Module to NoiseModule, so it's a bit more distinctive.

**Comment by bjz** [comment]
Here is the part of the Cargo docs relating to features: http://doc.crates.io/manifest.html#the-features-section

Still not happy with the `NoiseModule` name, but let's just merge this. I'll make some issues for the other things.
